### PR TITLE
added cluster restore

### PR DESCRIPTION
**Pull request by britztopher**
added tests and refactored restore in rdscluster file
added docs and sad paths

@kapilt @ewbankkit @mandeepbal I added the restore functionality to be able to restore from the latest snapshot if a snapshot isnt supplied. Take a look and be gentle :) this is my first python program. Included tests and docs on restore. Please let me know what else i can do, otherwise, happy peer reviewing 👍

### CODE REVIEW

File: c7n/resources/rdscluster.py [hunk 1]
```diff
@@ -216,5 +211,141 @@ def process(self, snapshots):
...
+
+        return x
+
+    def create_db_instances_within_cluster(self, snapshot):
+
+        db_instances = self.data.get('db-instances') or {}
+        num_of_instances = db_instances.get('number', 1)
+        db_identifier = db_instances.get('identifier', self.get_default_identifier(snapshot['DBClusterIdentifier']))
+        db_size = db_instances.get('size', 'db.r3.large')
+        post_fix = db_instances.get('post-fix', 'dev')
```
Comment (kapilt) [inline c7n/resources/rdscluster.py]: s/post-fix/id-suffix

File: c7n/resources/rdscluster.py [hunk 2]
```diff
@@ -216,5 +211,141 @@ def process(self, snapshots):
...
+
+        if db_identifier is None:
+            raise ClientError('MUST PROVIDE A db_identifier IN YOUR POLICY: Please delete cluster before trying again')
+        if db_size is None:
+            self.log_default_value('size:: default size is :: %s , which is minimum size', db_size)
+        if num_of_instances is None:
+            self.log_default_value('number:: number of db instances which is: %s', num_of_instances)
+
+        for instance in xrange(0, num_of_instances):
```
Comment (kapilt) [inline c7n/resources/rdscluster.py]: s/xrange/range for any sizes where dealing with in resources, the difference is in consequential, and it removes a py3 incompatibilty.

File: c7n/resources/rdscluster.py [hunk 3]
```diff
@@ -216,5 +211,141 @@ def process(self, snapshots):
...
+    schema = type_schema(
+        'restore', vpc_sec_groups={'type': 'array'}, port={'type': 'number'}, db_subnet_group_name={'type': 'string'},
+        snapshot_identifier={'type': 'string'})
+
+    def process(self, resources):
+
+        vpc_sec_groups = self.data.get('vpc-security-groups')
+        port = self.data.get('port')
+        db_subnet_group_name = self.data.get('db-security-group-name')
+        snapshot_id = self.data.get('snapshot-identifier');
```
Comment (kapilt) [inline c7n/resources/rdscluster.py]: this seems a little confusing in conjunction with the loop below, so if the snapshot is specified, then your restoring the database N times, where n is the number of snapshots in the account.

File: c7n/resources/rdscluster.py [hunk 4]
```diff
@@ -216,5 +211,141 @@ def process(self, snapshots):
...
+
+    """
+      Right now im only returning the first value returned by the cli. I sort them from bottom to top then append the top
+      value to the array then return the array.  I am setting this function up to handle an array of different
+      cluster's snapshot, but for right now I just want to get the single cluster restore working then Ill focus on
+      creating multiple rds cluster instances that have same values within an array in the policy.
+    """
+    @staticmethod
+    def get_latest_snapshot(snapshots):
+
```
Comment (kapilt) [inline c7n/resources/rdscluster.py]: This method seems suspect, its depending on result ordering from the aws api, which is effectively undefined. better would be to explicitly construct a map of identifiers to arrays of snapshots and return the map.

File: c7n/resources/rdscluster.py [hunk 5]
```diff
@@ -216,5 +211,141 @@ def process(self, snapshots):
...
+            try:
+                client.create_db_instance(**params)
+            except ClientError as e:
+                raise
+
+    """
+    TODO: Need to delete cluster when failure on database instance creation.  Wondering how I can call the Delete class
+    from within this function
+    """
```
Comment (kapilt) [inline c7n/resources/rdscluster.py]: in this context you can just use the api on the session, the Delete Action can be reused but it needs construction ala Action(config_dict, resource_manager_instance).process(resources)

File: c7n/resources/rdscluster.py (hunk 4 above)
Comment (britztopher) [inline c7n/resources/rdscluster.py]: the returned snapshots are actually in order of creation time

File: c7n/resources/rdscluster.py (hunk 1 above)
Comment (britztopher) [inline c7n/resources/rdscluster.py]: is this a regex for something?

File: c7n/resources/rdscluster.py (hunk 4 above)
Comment (kapilt) [inline c7n/resources/rdscluster.py]: right, but there going to be interleaved across multiple db instances.

File: c7n/resources/rdscluster.py (hunk 2 above)
Comment (britztopher) [inline c7n/resources/rdscluster.py]: so use range instead?  I dont know what you mean by `s/xrange/range` is that regex language for replace the string xrange with range?

File: c7n/resources/rdscluster.py (hunk 5 above)
Comment (britztopher) [inline c7n/resources/rdscluster.py]: Yeah i was planning on reusing the existing rds cluster delete, however, I havent figured out how to call it from my Restore class.

File: c7n/resources/rdscluster.py (hunk 3 above)
Comment (britztopher) [inline c7n/resources/rdscluster.py]: Exactly. When the user supplies the number of instances that is to be made within the cluster it loops through creating each one.

File: c7n/resources/rdscluster.py [hunk 6]
```diff
@@ -216,5 +211,141 @@ def process(self, snapshots):
...
+            if vpc_sec_groups is None:
+                self.log_default_value('vpc_sec_group')
+            elif db_subnet_group_name is None:
+                self.log_default_value('db_subnet_group_name')
+            elif port is None:
+                self.log_default_value('port')
+            else:
+                log.info('USING PARAMS PROVIDED IN POLICY')
+
+            print 'USING SNAPSHOT FOR CLUSTER::', latest_snapshot['DBClusterIdentifier']
```
Comment (britztopher) [inline c7n/resources/rdscluster.py]: yeah, sorry i was debugging something and log.debug or log.info wasnt printing in the format i was looking for.  Just forgot to remove this print statement

File: c7n/resources/rdscluster.py (hunk 1 above)
Comment (kapilt) [inline c7n/resources/rdscluster.py]: doh.. yeah its a short hand suggestion for a string replacement.

File: c7n/resources/rdscluster.py (hunk 2 above)
Comment (kapilt) [inline c7n/resources/rdscluster.py]: sorry, shorthand syntax, yes the intent was to suggest replacing the xrange string with range.

File: c7n/resources/rdscluster.py (hunk 4 above)
Comment (britztopher) [inline c7n/resources/rdscluster.py]: yes but right now you can only restore one cluster for one policy leaving it only to iterate over one rds cluster.  Meaning there is only ever going to be one DBClusterIdentifier.

File: tests/test_rdscluster.py [hunk 7]
```diff
@@ -145,3 +147,47 @@ def test_rdscluster_snapshot_trim(self):
...
         resources = p.run()
         self.assertEqual(len(resources), 2)
+
+
+class RDSClusterSnapshotRestore(BaseTest):
+
+    data = []
+
+    def test_get_latest_snapshot(self):
+        with open('./tests/data/placebo/test_rdscluster_snapshot_restore/rds.DescribeDBClusterSnapshots_1.json', 'r') as f:
```
Comment (kapilt) [inline tests/test_rdscluster.py]: from common import load_data
data = load_data('placebo/test_rdscluster_snapshot_restore/rds.DescribeDBClusterSnapshots_1.json')

**Comment by kapilt** [comment]
any updates on this?

**Comment by britztopher** [comment]
@kapilt sorry i got bogged down, ive done all you asked besides deletion as I need help with it.

**Comment by kapilt** [comment]
can you push the current branch so i can have a look then? thanks

**Comment by britztopher** [comment]
@kapilt  checked in latest in dev progress

File: c7n/resources/rdscluster.py [hunk 8]
```diff
@@ -72,30 +72,13 @@ def process(self, clusters):
                         'Deleted RDS instance: %s',
                         instance['DBInstanceIdentifier'])
 
-            params = {'DBClusterIdentifier': cluster['DBClusterIdentifier']}
```
Comment (kapilt) [inline c7n/resources/rdscluster.py]: why was this removed? afaics the `delete` action is no longer functional with this change

File: c7n/resources/rdscluster.py (hunk 3 above)
Comment (kapilt) [inline c7n/resources/rdscluster.py]: the issue is what if you have multiple snapshots for the same database?

File: c7n/resources/rdscluster.py (hunk 4 above)
Comment (kapilt) [inline c7n/resources/rdscluster.py]: where is that enforced? an action ideally shouldn't assume extant filters, if it does then that needs to be explicit in the validate() method for the action.

**Comment by kapilt** [comment]
superceded in part by #1224
